src/datasets/iterable_dataset.py

```diff
@@ -206,12 +206,21 @@ def __iter__(self):
                 function_args = [inputs] if self.input_columns is None else [inputs[col] for col in self.input_columns]
                 if self.with_indices:
                     function_args.append([current_idx + i for i in range(len(key_examples_list))])
+
                 transformed_batch = dict(batch)  # this will be updated with the function output
-                transformed_batch.update(self.function(*function_args))
+
+                processed_batch = self.function(*function_args)
                 # then remove the unwanted columns
-                if self.remove_columns:
+                if isinstance(self.remove_columns, Iterable):
                     for c in self.remove_columns:
-                        del transformed_batch[c]
+                        if c in transformed_batch:
+                            del transformed_batch[c]
```

Comment:
This is the sidefix: remove has to be done on input, and then merged with processed_input similarly to original map function.